File: game/src/main/scala/falcun/net/api/gui/animations/AnimationTimer.scala
package falcun.net.api.gui.animations

final class AnimationTimer(var start: Int, var end: Int, var duration: Long) {
  creation = System.currentTimeMillis
  var creation = 0L

  def getProgress: Int = {
    val dif = (System.currentTimeMillis - creation).toDouble
    var percent = dif / duration
    if (percent > 1) percent = 1
    val gap = end - start
    val progress = (percent * gap).toInt
    start + progress
  }

  def isDone: Boolean = System.currentTimeMillis > creation + duration

  def swapDirection(): Unit = {
    val now = System.currentTimeMillis
    var remaingTime = duration + creation - now
    if (remaingTime < 0) remaingTime = 0
    creation = now - remaingTime
    val s = start
    start = end
    end = s
  }
}

File: client/projects/Forge/src/main/java/falcun/net/api/oldgui/animations/AnimationTimer.java
package falcun.net.api.oldgui.animations;

public final class AnimationTimer {
	public int start;
	public int end;
	public long duration;
	public long creation;

	public AnimationTimer(int start, int end, long duration) {
		this.start = start;
		this.end = end;
		this.duration = duration;
		creation = System.currentTimeMillis();
	}

	public int getProgress() {
		double dif = (double) (System.currentTimeMillis() - creation);
		double percent = dif / duration;
		if (percent > 1) {
			percent = 1;
		}
		int gap = end - start;
		int progress = (int) (percent * gap);
		return start + progress;

	}

	public boolean isDone() {
		return System.currentTimeMillis() > creation + duration;
	}

	public void swapDirection() {
		long now = System.currentTimeMillis();
		long remaingTime = duration + creation - now;
		if(remaingTime < 0) {
			remaingTime = 0;
		}
		creation = now - remaingTime;
		int s = start;
		start = end;
		end = s;
	}

}